homeassistant/components/matter/update.py

```diff
@@ -0,0 +1,221 @@
+from homeassistant.helpers.restore_state import ExtraStoredData
+
+from .entity import MatterEntity
+from .helpers import get_matter
+from .models import MatterDiscoverySchema
+
+SCAN_INTERVAL = timedelta(hours=12)
+POLL_AFTER_INSTALL = 10
+
+ATTR_SOFTWARE_UPDATE = "software_update"
+
+
+@dataclass
+class MatterUpdateExtraStoredData(ExtraStoredData):
+    """Extra stored data for Z-Wave node firmware update entity."""
+
+    software_update: MatterSoftwareVersion | None = None
+
+    def as_dict(self) -> dict[str, Any]:
+        """Return a dict representation of the extra data."""
+        return {
+            ATTR_SOFTWARE_UPDATE: self.software_update,
+        }
+
+    @classmethod
+    def from_dict(cls, data: dict[str, Any]) -> MatterUpdateExtraStoredData:
+        """Initialize the extra data from a dict."""
+        return cls(data[ATTR_SOFTWARE_UPDATE])
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Matter lock from Config Entry."""
+    matter = get_matter(hass)
+    matter.register_platform_handler(Platform.UPDATE, async_add_entities)
+
+
+class MatterUpdate(MatterEntity, UpdateEntity):
+    """Representation of a Matter node capable of updating."""
+
+    _software_update: MatterSoftwareVersion | None = None
+    _cancel_update: CALLBACK_TYPE | None = None
+
+    @callback
+    def _update_from_device(self) -> None:
+        """Update from device."""
+
+        self._attr_installed_version = self.get_matter_attribute_value(
+            clusters.BasicInformation.Attributes.SoftwareVersionString
+        )
+
+        if self.get_matter_attribute_value(
+            clusters.OtaSoftwareUpdateRequestor.Attributes.UpdatePossible
+        ):
+            self._attr_supported_features = (
+                UpdateEntityFeature.INSTALL
+                | UpdateEntityFeature.PROGRESS
+                | UpdateEntityFeature.SPECIFIC_VERSION
+            )
+
+        update_state: clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum = (
+            self.get_matter_attribute_value(
+                clusters.OtaSoftwareUpdateRequestor.Attributes.UpdateState
+            )
+        )
+        if (
+            update_state
+            == clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum.kIdle
+        ):
+            self._attr_in_progress = False
+            return
+
+        update_progress: int = self.get_matter_attribute_value(
+            clusters.OtaSoftwareUpdateRequestor.Attributes.UpdateStateProgress
+        )
+
+        if (
+            update_state
+            == clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum.kDownloading
+            and update_progress is not None
+            and update_progress > 0
+        ):
+            self._attr_in_progress = update_progress
+        else:
+            self._attr_in_progress = True
+
+    async def async_update(self) -> None:
```

Comment:
Why does this method not just call `self._update_from_device` ?

Reply:
`_update_from_device` gets called on every attribute update as pushed by the device. We don't want to trigger a update check whenever a random attribute changes on the device.